libmamba/src/core/package_handling.cpp

```diff
@@ -57,7 +58,104 @@ namespace mamba
         const fs::u8path& m_file;
     };
 
-    static int copy_data(archive* ar, archive* aw)
+    class libarchive_read_raii
+    {
+    public:
+        libarchive_read_raii()
+            : libarchive_read_raii(archive_read_new()){};
+        static libarchive_read_raii read_disk()
+        {
+            return libarchive_read_raii(archive_read_disk_new());
+        }
+
+        ~libarchive_read_raii()
+        {
+            archive_read_free(m_archive);
+        }
+
+        operator archive*()
+        {
+            return m_archive;
+        }
+
+    private:
+        libarchive_read_raii(archive* a)
+            : m_archive(a)
+        {
+            if (!m_archive)
+            {
+                throw std::runtime_error("Could not create libarchive read object");
+            }
+        }
+
+        archive* m_archive;
+    };
+
+    class libarchive_entry_raii
+    {
+    public:
+        libarchive_entry_raii()
+            : m_entry(archive_entry_new())
+        {
+            if (!m_entry)
+            {
+                throw std::runtime_error("Could not create libarchive entry object");
+            }
+        }
+
+        ~libarchive_entry_raii()
+        {
+            archive_entry_free(m_entry);
+        }
+
+        operator archive_entry*()
+        {
+            return m_entry;
+        }
+
+    private:
+        archive_entry* m_entry;
+    };
+
+    class libarchive_write_raii
+    {
+    public:
+        libarchive_write_raii()
+            : libarchive_write_raii(archive_write_new())
+        {
+        }
+
+        static libarchive_write_raii write_disk()
+        {
+            return libarchive_write_raii(archive_write_disk_new());
+        }
+
+        ~libarchive_write_raii()
+        {
+            archive_write_free(m_archive);
+        }
+
+        operator archive*()
+        {
+            return m_archive;
+        }
+
+    private:
+        libarchive_write_raii(archive* a)
+            : m_archive(a)
+        {
+            if (!m_archive)
+            {
+                throw std::runtime_error("Could not create libarchive write object");
+            }
+        }
+
+        archive* m_archive;
+    };
+
+    void stream_extract_archive(libarchive_read_raii& a, const fs::u8path& destination);
+
+    static int copy_data(libarchive_read_raii& ar, libarchive_write_raii& aw)
```

Comment:
I'm surprised that they are both modified?

Reply:
yes, e.g. the `a->error` is set on the archive if either read or write fails.